// falkordb/dist/src/clients/nullClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullClient = void 0;
/**
 * The `NullClient` class is a placeholder implementation of the `Client` interface.
 *
 * This class is designed to be used in scenarios where a client is required, but no actual
 * implementation is available. All methods in this class throw a "Method not implemented."
 * error, indicating that the functionality has not been provided.
 *
 * The `NullClient` can serve as a base class or a stub for future implementations, or as a
 * fallback in cases where a functional client is not needed or cannot be instantiated.
 *
 */
class NullClient {
    getConnection() {
        throw new Error("Method not implemented.");
    }
    init(_falkordb) {
        throw new Error("Method not implemented.");
    }
    list() {
        throw new Error("Method not implemented.");
    }
    configGet(_configKey) {
        throw new Error("Method not implemented.");
    }
    configSet(_configKey, _value) {
        throw new Error("Method not implemented.");
    }
    info(_section) {
        throw new Error("Method not implemented.");
    }
    query(_graph, _query, _options) {
        throw new Error("Method not implemented.");
    }
    profile(_graph, _query) {
        throw new Error("Method not implemented.");
    }
    roQuery(_graph, _query, _options) {
        throw new Error("Method not implemented.");
    }
    copy(_srcGraph, _destGraph) {
        throw new Error("Method not implemented.");
    }
    delete(_graph) {
        throw new Error("Method not implemented.");
    }
    explain(_graph, _query) {
        throw new Error("Method not implemented.");
    }
    slowLog(_graph) {
        throw new Error("Method not implemented.");
    }
    memoryUsage(_graph, _options) {
        throw new Error("Method not implemented.");
    }
    constraintCreate(_graph, _constraintType, _entityType, _label, ..._properties) {
        throw new Error("Method not implemented.");
    }
    constraintDrop(_graph, _constraintType, _entityType, _label, ..._properties) {
        throw new Error("Method not implemented.");
    }
    udfLoad(_name, _script, _replace) {
        throw new Error("Method not implemented.");
    }
    udfList(_lib, _withCode) {
        throw new Error("Method not implemented.");
    }
    udfFlush() {
        throw new Error("Method not implemented.");
    }
    udfDelete(_lib) {
        throw new Error("Method not implemented.");
    }
    quit() {
        throw new Error("Method not implemented.");
    }
    disconnect() {
        throw new Error("Method not implemented.");
    }
}
exports.NullClient = NullClient;
