// falkordb/dist/src/clients/single.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Single = void 0;
class Single {
    client;
    constructor(client) {
        this.client = client;
    }
    init(_falkordb) {
        return Promise.resolve();
    }
    async query(graph, query, options, compact = true) {
        return await this.client.falkordb.query(graph, query, options, compact);
    }
    async roQuery(graph, query, options, compact = true) {
        return await this.client.falkordb.roQuery(graph, query, options, compact);
    }
    async delete(graph) {
        const reply = this.client.falkordb.delete(graph);
        return reply.then(() => { });
    }
    async explain(graph, query) {
        return this.client.falkordb.explain(graph, query);
    }
    async profile(graph, query) {
        return this.client.falkordb.profile(graph, query);
    }
    async list() {
        return this.client.falkordb.list();
    }
    async configGet(configKey) {
        return this.client.falkordb.configGet(configKey);
    }
    async configSet(configKey, value) {
        const reply = this.client.falkordb.configSet(configKey, value);
        return reply.then(() => { });
    }
    async info(section) {
        return this.client.falkordb.info(section);
    }
    async slowLog(graph) {
        return this.client.falkordb.slowLog(graph);
    }
    async memoryUsage(graph, options) {
        return this.client.falkordb.memoryUsage(graph, options);
    }
    async constraintCreate(graph, constraintType, entityType, label, ...properties) {
        const reply = this.client.falkordb.constraintCreate(graph, constraintType, entityType, label, ...properties);
        return reply.then(() => { });
    }
    async constraintDrop(graph, constraintType, entityType, label, ...properties) {
        const reply = this.client.falkordb.constraintDrop(graph, constraintType, entityType, label, ...properties);
        return reply.then(() => { });
    }
    async udfLoad(name, script, replace = false) {
        return this.client.falkordb.udfLoad(name, script, replace);
    }
    async udfList(lib, withCode = false) {
        return this.client.falkordb.udfList(lib, withCode);
    }
    async udfFlush() {
        return this.client.falkordb.udfFlush();
    }
    async udfDelete(lib) {
        return this.client.falkordb.udfDelete(lib);
    }
    async copy(srcGraph, destGraph) {
        return this.client.falkordb.copy(srcGraph, destGraph);
    }
    quit() {
        return this.disconnect();
    }
    async disconnect() {
        const reply = this.client.disconnect();
        return reply.then(() => { });
    }
    async getConnection() {
        return this.client;
    }
}
exports.Single = Single;
