// falkordb/dist/src/commands/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushQueryArguments = pushQueryArguments;
const CONFIG_GET = require("./CONFIG_GET");
const CONFIG_SET = require("./CONFIG_SET");
const DELETE = require("./DELETE");
const EXPLAIN = require("./EXPLAIN");
const INFO = require("./INFO");
const LIST = require("./LIST");
const PROFILE = require("./PROFILE");
const QUERY = require("./QUERY");
const RO_QUERY = require("./RO_QUERY");
const SLOWLOG = require("./SLOWLOG");
const CONSTRAINT_CREATE = require("./CONSTRAINT_CREATE");
const CONSTRAINT_DROP = require("./CONSTRAINT_DROP");
const COPY = require("./COPY");
const SENTINEL_MASTER = require("./SENTINEL_MASTER");
const SENTINEL_MASTERS = require("./SENTINEL_MASTERS");
const MEMORY_USAGE = require("./MEMORY_USAGE");
const UDF_LOAD = require("./UDF_LOAD");
const UDF_LIST = require("./UDF_LIST");
const UDF_FLUSH = require("./UDF_FLUSH");
const UDF_DELETE = require("./UDF_DELETE");
exports.default = {
    CONFIG_GET,
    configGet: CONFIG_GET,
    CONFIG_SET,
    configSet: CONFIG_SET,
    DELETE,
    delete: DELETE,
    EXPLAIN,
    explain: EXPLAIN,
    INFO,
    info: INFO,
    LIST,
    list: LIST,
    PROFILE,
    profile: PROFILE,
    QUERY,
    query: QUERY,
    RO_QUERY,
    roQuery: RO_QUERY,
    SLOWLOG,
    slowLog: SLOWLOG,
    CONSTRAINT_CREATE,
    constraintCreate: CONSTRAINT_CREATE,
    CONSTRAINT_DROP,
    constraintDrop: CONSTRAINT_DROP,
    COPY,
    copy: COPY,
    SENTINEL_MASTER,
    sentinelMaster: SENTINEL_MASTER,
    SENTINEL_MASTERS,
    sentinelMasters: SENTINEL_MASTERS,
    MEMORY_USAGE,
    memoryUsage: MEMORY_USAGE,
    UDF_LOAD,
    udfLoad: UDF_LOAD,
    UDF_LIST,
    udfList: UDF_LIST,
    UDF_FLUSH,
    udfFlush: UDF_FLUSH,
    UDF_DELETE,
    udfDelete: UDF_DELETE,
};
function pushQueryArguments(args, graph, query, options, compact) {
    args.push(graph);
    if (typeof options === 'number') {
        args.push(query);
        pushTimeout(args, options);
    }
    else {
        args.push(options?.params ?
            `CYPHER ${queryParamsToString(options.params)} ${query}` :
            query);
        if (options?.TIMEOUT !== undefined) {
            pushTimeout(args, options.TIMEOUT);
        }
    }
    if (compact) {
        args.push('--compact');
    }
    return args;
}
function pushTimeout(args, timeout) {
    args.push('TIMEOUT', timeout.toString());
}
function queryParamsToString(params) {
    const parts = [];
    for (const [key, value] of Object.entries(params)) {
        parts.push(`${key}=${queryParamToString(value)}`);
    }
    return parts.join(' ');
}
function queryParamToString(param) {
    if (param === null) {
        return 'null';
    }
    switch (typeof param) {
        case 'string':
            return `"${param.replace(/["\\]/g, '\\$&')}"`;
        case 'number':
        case 'boolean':
            return param.toString();
    }
    if (Array.isArray(param)) {
        return `[${param.map(queryParamToString).join(',')}]`;
    }
    else if (typeof param === 'object') {
        const body = [];
        for (const [key, value] of Object.entries(param)) {
            body.push(`${key}:${queryParamToString(value)}`);
        }
        return `{${body.join(',')}}`;
    }
    else {
        throw new TypeError(`Unexpected param type ${typeof param} ${param}`);
    }
}
